src/clients/blocks.ts

```diff
@@ -1,49 +1,65 @@
-import { gql, GraphQLClient } from 'graphql-request';
 import _ from 'lodash';
 
+import { DBClient, Table } from '../db/db';
+import { Block } from '../types/block';
 import { ChainId } from '../types/chain';
 
 import { EVMClient } from './evm';
 
-const blocksAPI = new GraphQLClient(process.env.BLOCKS_SUBGRAPH_ENDPOINT!);
 
 export type BlocksClient = {
-  fetchBlockTimestamps: (chainId: ChainId, blockNumbers: string[]) => Promise<{ [blockNumber: string]: string }>;
+  fetchBlockTimestamps: (chainId: ChainId, blockNumbers: string[]) => Promise<{ 
+    timestampsByBlockNumber: { [blockNumber: string]: string },
+    newBlocks: Block[] 
+  }>;
 }
 
-const init = async (evmClients: { [chainId in ChainId]: EVMClient }) => {
+const init = async (
+  evmClients: { [chainId in ChainId]: EVMClient },
+  dbClient: DBClient
+) => {
   return {
     fetchBlockTimestamps: async (
       chainId: ChainId,
       blockNumbers: string[],
-    ): Promise<{ [blockNumber: string]: string }> => {
-      let result = {};
-
-      // only using the graph for Ethereum. All other EVM chains will use RPC calls
-      if (chainId === ChainId.ethereum){
-        const blockNumbersForQuery = JSON.stringify(blockNumbers);
-        const { blocks } = await blocksAPI.request(
-          gql`
-          {
-            blocks(first:${blockNumbers.length}, where: {number_in: ${blockNumbersForQuery}}) {
-              number
-              timestamp
-            }
-          }
-          `,
-        );
-        
-        const responseByBlockNumber = _.keyBy(blocks, 'number');
-        const blockTimes = _.mapValues(responseByBlockNumber, response => response.timestamp)
-        result = blockTimes;
-      } else {
-        const blockNumbersParsed: number[] = blockNumbers.map(blockNumber => parseInt(blockNumber))
-
-        const blocks = await evmClients[chainId].getBlockTimestampsByBlockNumber(blockNumbersParsed);
-        result = blocks
+    ): Promise<{ 
+      timestampsByBlockNumber: { [blockNumber: string]: string },
+      newBlocks: Block[] 
+    }> => {
+      let timestampsByBlockNumber: { [blockNumber: string]: string } = {};
+
+      const existingBlocks = await dbClient.getRecords<Block>(Table.blocks, [
+        ['where', ['chainId', chainId]],
+        ['whereIn', ['blockNumber', blockNumbers]]
+      ])
```

Comment:
hm so there will usually be 1000 or so blocks in this whereIn query? I guess that should be fine? as long as query pretty much instant?

Reply:
we got 120k transfers, so first run probably 120 loops of this processor

Reply:
Yeah usually 100 blocks because we slice it to that length. This doesn't change how many times the processor will run, just how many requests will be sent. The queries are pretty much instant, and after this has run a few times the number of requests drops significantly as we don't fetch duplicate blocks and just read those from the DB